text/0000-build-config.md

```diff
@@ -0,0 +1,87 @@
+# Meta
+[meta]: #meta
+- Name: Build config
+- Start Date: 2022-08-29
+- Author(s): [samj1912](https://github.com/samj1912)
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes an easy way to configure build images to allow specifying a `/cnb/build-config/env` CNB environment directory that allows updating the Buildpack `detect` and `build` environment based on the directory.
+
+
+# Definitions
+[definitions]: #definitions
+
+- CNB environment directory: A directory that follows the conventions as defined [here](https://github.com/buildpacks/spec/blob/main/buildpack.md#provided-by-the-buildpacks)
+- Operator: Owner of builders and stacks. See [CNB Operator guide](https://buildpacks.io/docs/operator-guide/).
+
+
+# Motivation
+[motivation]: #motivation
+
+Often times, especially in enterprise settings, organizations often have to update the buildpacks to use internal mirrors, proxies and other settings which can be configured easily with environment variables.
+
+Some examples include -
+- `GOPROXY`
+- `PIP_INDEX_URL`
+- `npm_config_registry`
+- `BUNDLE_MIRROR__URL`
+
+The buildpack logic in the Buildpacks largely remains the same, except these environment variables might need to be injected during the `build` and `detect` phases.
+
+The environment variables may ideally also take precendence over any user provided values to ensure that the operators have full control over their builders.
```

Comment:
hmm, this feels unexpected

Reply:
This mostly provides a choice to operators. If they don't want to have a higher precedence than user provided variables, they can choose to use `VAR_NAME.default`. If they do want a higher precedence they can choose to use `VAR_NAME.override`. The reason this needs to be at a higher precedence level than the user is to allow modifying user args to conform to any buildpack security/compliance requirements. For eg. being able to prepend/append args to user provided ones or being able to override user-provided invalid arguments for variables like proxies/mirrors etc. if desired. It always allows them to give higher precedence to user provided args via `.DEFAULT`. Buildpacks are already able to do this via setting the `clear-env` flag in their buildpack.toml or choosing to override/modify input user args.

Reply:
Another alternative may be to introduce a `env.pre` and `env.post` folder to allow for both behaviors if that is important. But for us, being able to sanitize/mutate user input is extremely important. Sort of similar to k8s mutating/validation webhooks.

Reply:
> Another alternative may be to introduce a env.pre and env.post folder to allow for both behaviors if that is important. But for us, being able to sanitize/mutate user input is extremely important. Sort of similar to k8s mutating/validation webhooks.

Would these folders be in `/cnb/platform`?

My main concern from this RFC, is "yet another way to configure the same thing". If this RFC is primarily focused on order of precedence and not portability (embedded in build image) then I'd prefer to see if we can adapt `/cnb/platform` to provide the desired behavior.

Reply:
@jromero it is both. The portability aspects of it are fairly important.

> Would these folders be in /cnb/platform? 

These would be folders in the build image.

Reply:
```suggestion
The environment variables may be able to take precedence over user-provided values, if configured to, which ensures that operators can take full control over their builders if that is required.
```

Rephrases this a bit. I agree the default behavior should be to have user-specified config take precedence, but I also know operators might want or need to have control as well.